### PR TITLE
Fix Cash App Pay token reuse issues (with subscriptions)

### LINKED ISSUE #3171: [CashApp] Support subscription with a free trial and changing a subscription's payment method to Cash App
### Describe the bug
In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3156, we added broad support for using Cash App Pay to purchase and renew subscriptions. 

While working on that integration, I discovered that Cash App cannot be used to purchase $0 subscriptions (eg free trials). This limitation also extended to changing a subscription's payment method to Cash App. 

The underlying issue that is leading to Cash App not supporting those Subscription features is the same. When you purchase a free trial subscription or attempt to change a subscription's payment method to Cash App Pay, the `process_payment_with_deferred_intent()` function [attempts to setup an intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8.3.0/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L790). If you are using a new Cash App payment method, this works just fine - setting up a Cash App intent is allowed. Where this fails is **if you use a saved cash app token**. In that case you will receive the following error. 

| Screenshot 2024-06-04 at 2 29 53 PM | Screenshot 2024-06-03 at 11 07 11 AM |
|--------|--------|

```
The provided Cash App Pay payment method is already reusable with an active mandate. You can't associate it with a new SetupIntent.
```
<p align="center"><sup>Full error message included for search-ability.</sup></p>

In order to release broad subscription support in the initial release of Cash App, we disabled Cash App Pay in this scenarios. This issue is in place to develop a fix for that to allow Cash App to be available for those features. 

### To Reproduce

**Free trial** 
1. Change your Stripe Plugin credentials to a US based account.
3. Change your store currency to USD.
4. Enable Cash App Pay from the Stripe payment methods screen.
5. Enable WooCommerce Subscriptions. 
6. Create a Subscription product with a free trial. 
7. Add the product to your cart. 
8. Go to the checkout
9. Notice that Cash App Pay isn't available.
10. Remove the `WC_Stripe_UPE_Payment_Method_Cash_App_Pay::is_enabled_at_checkout()` function.
    - It is the function which enforces this $0 payment restriction.  
11. Refresh the checkout. Cash App should be available now.
13. Purchase the subscription using Cash App - this will create a token. 
14. Add the free trial product to your cart again. 
15. Attempt to complete the sign up using the saved payment method.
16. You should received the error message mentioned above. 

**Changing payment method** 
1. Purchase a subscription product using a card.
1. Go to the **My Account → Subscriptions**
2. View your active subscription. 
3. Click the "Change payment" button.
4. On the **Change payment method** page notice that Cash App isn't available. 
5. Remove the code in the `WC_Stripe_UPE_Payment_Method_Cash_App_Pay` constructor which removes the `subscription_payment_method_change_customer` support flag. 
6. Refresh the Change payment method page. Cash App should be available now.
7. Attempt to change the subscription's payment to **a saved cash app token**. 
8. You should receive the error I mentioned above. 

### Expected behavior
- You should be able to purchase free trial products with saved Cash App Pay tokens. 
- You should be able to change a subscription's payment method to a saved Cash App Pay token.

### Additional context

At its core the issue here is that Stripe doesn't allow you to call setup intent and pass a saved Cash App token. I'm not sure why that is because SEPA and card tokens work just fine. With this in mind we will need to find a work around. Either: 

1. Don't send the setup intent when using a saved Cash App token. This complicates things as it splits the flow of `process_payment_with_deferred_intent()`. We would need to save the tokens etc and then exit the function without sending a request to Stripe. 
2. Fetch the set up intent that was originally used to create that Cash App token and process it as if it was used to process this new order's set up. 
3. Another option.

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes #3171

## Changes proposed in this Pull Request:

This PR enables the use of Cash App Pay when paying for subscriptions with free trials and as an option when changing a subscription's payment method.

For that, I have updated the usage of the Cash App Pay confirm method. The wallet hash generation method includes a new parameter (the payment intent type).

**Known issue**: The subscription extension adds a "success message" before the redirect is completed, which can be confusing for customers.

## Testing instructions

- Checkout to this branch on your local environment (`fix/cashapp-token-reuse2`)
- Run `npm install`, `npm run build:webpack` and `npm run up`

(the following testing guide was extracted from the [original issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3171))

**Free trial** 
1. Change your Stripe Plugin credentials to a US-based account.
3. Change your store currency to USD.
4. Enable Cash App Pay from the Stripe payment methods screen.
5. Enable WooCommerce Subscriptions. 
6. Create a Subscription product with a free trial. 
7. Add the product to your cart. 
8. Go to the checkout
9. Confirm that you can select Cash App Pay as a payment method
10. Purchase the subscription using Cash App - this will create a token. 
11. Add the free trial product to your cart again. 
12. Confirm that you can complete the sign-up using the saved payment method.

**Changing payment method** 
1. Purchase a subscription product using a card.
1. Go to the **My Account → Subscriptions**
2. View your active subscription. 
3. Click the "Change payment" button.
6. Confirm that the Cash App Pay method is now available.
7. Confirm that you can change the subscription's payment to **a saved cash app token**. 

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 1]
```diff
@@ -834,6 +834,14 @@ private function process_payment_with_deferred_intent( int $order_id ) {
 
 				// Create a payment intent, or update an existing one associated with the order.
 				$payment_intent = $this->process_payment_intent_for_order( $order, $payment_information );
+			} elseif ( $payment_information['is_using_saved_payment_method'] && 'cashapp' === $selected_payment_type ) {
+				// If the payment method is Cash App Pay and the order has no cost, mark the order as paid.
+				$order->payment_complete();
+
+				return [
+					'result'   => 'success',
+					'redirect' => '',
```
Comment (james-allan) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]:
```suggestion
					'redirect' => $this->get_return_url( $order ),
```

Without a redirect, using a saved cash app token for an order that doesn't need payment, just leaves the customer on the checkout page.

**Review by james-allan — CHANGES_REQUESTED** [review]
Hey @wjrosa. I've been following some general Subscription testing instructions. I've listed them all below. I ran into a couple of issues, I'll follow up and finish testing shortcode checkouts tomorrow or later this week. 

**BLOCK CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [x] New token
   - [ ] ❗ Existing token
       - Selecting an existing token and attempting to sign up to a free trial ~~just leaves the customer on the checkout page~~ leaves the subscription without a stored payment method. There's more to this as well. I'll leave a comment below. *
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**SHORTCODE CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [ ] New token
   - [ ] Existing token
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**OTHER FEATURES**
- **Renew subscriptions**
   - [x] New token
   - [x] Existing token
- **Change subscription payment**
   - ~~❗ I cannot get Cash App to show as an option on the change payment method screen. ([screenshot](https://github.com/user-attachments/assets/07d9075c-00a9-43b3-8f19-22a4d7d172b6))~~
   - [x] New token
   - [x] Existing token
- [x] Add payment method via My Account.
   - ~~❗ The Cash App modal isn't shown. There's a browser console error and [this error](https://github.com/user-attachments/assets/4bb26dac-b1e0-45be-b3f5-76d262874c00) in Stripe.~~ This is not related to this PR. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3311
- **Failed/manual renewal - Block Checkout**
   - [x] New token
   - [x] Existing token
- **Failed/manual renewal - Shortcode Checkout**
   - [x] New token
   - [x] Existing token

\* **Free trial - existing token issue.** Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the `process_payment_with_deferred_intent()` function that things like saving the payment method information doesn't happen.

File: includes/class-wc-stripe-intent-controller.php [hunk 2]
```diff
@@ -956,7 +956,7 @@ public function is_mandate_data_required( $selected_payment_type, $is_using_save
 	public function create_and_confirm_setup_intent( $payment_information ) {
 		$request = [
 			'payment_method'       => $payment_information['payment_method'],
-			'payment_method_types' => $payment_information['payment_method_types'],
+			'payment_method_types' => $payment_information['payment_method_types'] ?? [ $payment_information['selected_payment_type'] ],
```
Comment (wjrosa) [inline includes/class-wc-stripe-intent-controller.php]: Hi @james-allan ! Your change [here](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/3929b33faa6aedf7864b531e4ce845ebdaa9f186) broke this part when adding Cash App Pay tokens. `$payment_information['payment_method_types']` does not exist in this context. I have added the previous version as a fallback to fix it.

**Comment by wjrosa** [comment]
@james-allan 
> * Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.

I believe the latest changes fixed this issue.

> Change subscription payment
❗ I cannot get Cash App to show as an option on the change payment method screen. ([screenshot](https://github.com/user-attachments/assets/07d9075c-00a9-43b3-8f19-22a4d7d172b6))

I had the same issue and did some debugging. The issue happens on `togglePaymentMethodForCountry`. For some reason `getStripeServerData()?.customerData?.billing_country` is not being populated. My country is set to US (on the checkout and on the Stripe side as a customer), but `customerData` is undefined.

It looks like this property is set by calling `useCustomerData`. Not sure if we need to call it manually on the change payment method page 🤔  (it is not called there). Still, does not seem directly related to this PR.

**Comment by wjrosa** [comment]
I actually fixed the above on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3263/commits/64b20d5edf7179fae8fda283b3b0f3e0fe865dd8 @james-allan 👍

File: includes/class-wc-stripe-intent-controller.php (hunk 2 above)
Comment (james-allan) [reply includes/class-wc-stripe-intent-controller.php]: 👍 Thanks for the heads up. I also changed it back to fallback to the `selected_payment_type` like you did. https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3311/commits/ebcd3b3604638c20673f6d59678159a0f780b86a.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 3]
```diff
@@ -1647,10 +1661,7 @@ public function set_payment_method_title_for_order( $order, $payment_method_type
 		$order->save();
 
 		// Update the subscription's purchased in this order with the payment method ID.
-		if ( isset( $this->payment_methods[ $payment_method_type ] ) ) {
```
Comment (wjrosa) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:1616]: This check is not necessary since we have the same thing above.

**Comment by wjrosa** [comment]
@james-allan I thought this was not working:

> Change subscription payment
> * New token

But it is. It seems that we have an issue with this cache:
```
self::PAYMENT_METHODS_TRANSIENT_KEY . $payment_method_type . $this->get_id()
```
It is not saving the new Cash App method there. Consequently, when displaying the method, it is shown as "N/A".

But it's not related to this change, I guess.

**Comment by wjrosa** [comment]
hi @james-allan ! I tested the other flows from your checklist myself, and I guess it is all good. One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process

**Comment by james-allan** [comment]
> > Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.

> I believe the latest changes fixed this issue.

@wjrosa this flow is causing an issue for me. If you add a subscription with a free trial to your cart and then use a saved Cash App token, the resulting subscription will have missing payment method token information.


<p align="center">
Screenshot 2024-08-13 at 2 53 17 PM
</br>
<sup>Empty payment method information on <strong>My Account → Subscriptions</strong> list.</sup>
</p> 

<p align="center">
Screenshot 2024-08-13 at 2 55 20 PM
</br>
<sup>Empty payment method information on <strong>WooCommerce → Subscriptions (edit)</strong> screen</sup>
</p> 

To fix this we'd need to update [this section](https://github.com/woocommerce/woocommerce-gateway-stripe/blame/fix/cashapp-token-reuse2/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L849-L856) of the code. If you purchase a free subscription with a saved token, the payment_complete and early return occurs without storing the token to the order. We essentially need to make sure we still save the token to the order too. re: 

```php
$this->maybe_update_source_on_subscription_order(
	$order,
	(object) [
		'payment_method' => $payment_information['payment_method'],
		'customer'       => $payment_information['customer'],
	],
	$this->get_upe_gateway_id_for_order( $upe_payment_method )
);
```

**Comment by james-allan** [comment]
> One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process

Yeah that seems a little off. Are you sure one of them is just going straight to active? My webhooks are so quickly processed that I don't notice any difference. But yes, if you're using a saved payment method or a new token to pay for a new subscription that has an upfront cost I'd expect both of them to involve a webhook process.

In my case when I use a new or existing token by the time I land on the order received page the subscription is active.

Now that I think on it, perhaps what you're seeing is the difference between a setup intent and just payment intent. 🤷 In my testing though both result in a `payment_intent.succeeded` and a `charge.succeeded` webhook.

**Review by james-allan — CHANGES_REQUESTED** [review]
Changes look pretty good @wjrosa. I just left a comment about the subscriptions with a free trial being purchased with a saved payment method.

**Comment by wjrosa** [comment]
> > > Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.
> 
> > I believe the latest changes fixed this issue.
> 
> @wjrosa this flow is causing an issue for me. If you add a subscription with a free trial to your cart and then use a saved Cash App token, the resulting subscription will have missing payment method token information.
> 
> Screenshot 2024-08-13 at 2 53 17 PM Empty payment method information on **My Account → Subscriptions** list.
> 
> Screenshot 2024-08-13 at 2 55 20 PM Empty payment method information on **WooCommerce → Subscriptions (edit)** screen
> 
> To fix this we'd need to update [this section](https://github.com/woocommerce/woocommerce-gateway-stripe/blame/fix/cashapp-token-reuse2/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L849-L856) of the code. If you purchase a free subscription with a saved token, the payment_complete and early return occurs without storing the token to the order. We essentially need to make sure we still save the token to the order too. re:
> 
> ```
> $this->maybe_update_source_on_subscription_order(
> 	$order,
> 	(object) [
> 		'payment_method' => $payment_information['payment_method'],
> 		'customer'       => $payment_information['customer'],
> 	],
> 	$this->get_upe_gateway_id_for_order( $upe_payment_method )
> );
> ```

Thanks, @james-allan ! I fixed this on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3263/commits/b8bd1afac2d8d885901ab3b0cdf87baeaa828211

**Comment by wjrosa** [comment]
> > One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process
> 
> Yeah that seems a little off. Are you sure one of them is just going straight to active? My webhooks are so quickly processed that I don't notice any difference. But yes, if you're using a saved payment method or a new token to pay for a new subscription that has an upfront cost I'd expect both of them to involve a webhook process.
> 
> In my case when I use a new or existing token by the time I land on the order received page the subscription is active.
> 
> Now that I think on it, perhaps what you're seeing is the difference between a setup intent and just payment intent. 🤷 In my testing though both result in a `payment_intent.succeeded` and a `charge.succeeded` webhook.

Any ideas where this status change to `active` is handled? Maybe that's something to do with the early return after calling `payment_complete()`.

**Comment by wjrosa** [comment]
Still not sure where the issue is, but I did some debugging and saw that:
- When purchasing a product using an existing Cash App Pay token, we are calling `create_and_confirm_payment_intent`
- So, we are indeed processing the payment intent already
- Then getting inside this condition:
```
if ( $payment_needed ) {
				// Use the last charge within the intent to proceed.
				$charge = $this->get_latest_charge_from_intent( $payment_intent );

				// Only process the response if it contains a charge object. Intents with no charge require further action like 3DS and will be processed later.
				if ( $charge ) {
					$this->process_response( $charge, $order );
				}
```
- The `$charge` value will be from the intent we just processed before
- As it is `captured` and not `pending` but `succeeded`, it will get into this condition:
```
			if ( 'succeeded' === $response->status ) {
```
- Making the order `completed`, and triggering the hook to change the subscription status along it (to `active`)

So, what should we do in this case? (when using a saved Cash App Pay token for a paid subscription)
1. Leave it as is, as the intent is indeed being processed, and the webhook call will probably ignore this order/subscription
2. Skip this line and also return early:
```
// Create a payment intent, or update an existing one associated with the order.
$payment_intent = $this->process_payment_intent_for_order($order, $payment_information);
```
(but it may break something as we will have only the setup intent when the order/subscription is processed by the webhook call)
3. Return early before this check:
```
			if ( $payment_needed ) {
				// Use the last charge within the intent to proceed.
				$charge = $this->get_latest_charge_from_intent( $payment_intent );

				// Only process the response if it contains a charge object. Intents with no charge require further action like 3DS and will be processed later.
				if ( $charge ) {
					$this->process_response( $charge, $order );
				}
```
4. Force it to fall into one of the next conditions if it makes sense: `elseif ( $this->is_changing_payment_method_for_subscription() )` or `elseif ( in_array( $payment_intent->status, self::SUCCESSFUL_INTENT_STATUS, true ) )` (by looking I would say only the last would make a bit sense)

**Comment by wjrosa** [comment]
@james-allan I did some additional analysis today for the status difference I mentioned earlier (after your fix for live payments), and I think the issue might be the `$order->payment_complete();`

### Non-saved token
- Order is kept as `pending`
- Confirmation is made on the frontend (QR code scan)

### Saved token
- Order is set to `completed` due to the call to `$order->payment_complete();`
- Nothing else is done

So:
- Should we have a QR code scan for saved tokens? From our previous conversations, I think it does not make sense (or did it change due to the live payments fix?)
- Should we just remove the call to `payment_complete` here instead?

**Comment by wjrosa** [comment]
I just tested this on a JN site, and the webhook call worked. So, maybe we don't need to do anything else here.

**Comment by james-allan** [comment]
> Should we have a QR code scan for saved tokens? From our previous conversations, I think it does not make sense (or did it change due to the live payments fix?)

Nah I don't think we do need to display the QR code because the PM has already been set up and we have a saved token. 

> I just tested this on a JN site, and the webhook call worked. So, maybe we don't need to do anything else here.

Ok I'll rerun through some tests and confirm this works as expected.

**Review by james-allan — APPROVED** [review]
**BLOCK CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [x] New token
   - [x] Existing token
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**OTHER FEATURES**
- **Renew subscriptions**
   - [x] New token
   - [x] Existing token
- **Change subscription payment**
   - [x] New token
   - [x] Existing token
- [x] Add payment method via My Account.
- **Failed/manual renewal - Block Checkout**
   - [x] New token
   - [x] Existing token

I also ran a basic test using a new token and a saved token on the shortcode checkout just to be sure.

**Pull request merged**